app/controllers/api/v1/messages_controller.rb

```diff
@@ -74,9 +69,18 @@ def mark_all_read
 
       private
 
+      def apply_filters(messages, options)
+        %i[ids offer_id order_id item_id].map do |f|
+          messages = messages.send("filter_by_#{f}", options[f]) if options[f]
+        end
+        messages = messages.with_state_for_user(current_user, options[:state].split(',')) if options[:state].present?
+        messages
+      end
+
       def apply_scope(records, scope)
         return records unless ALLOWED_SCOPES.include? scope
-        records.where("messages.#{scope}_id IS NOT NULL")
+
+        records.where("messages.messageable_type = '#{scope.camelize}'")
```

Comment:
Potential SQL injection, can be sanitized with

`records.where("messages.messageable_type = (?)", scope.camelize)`